src/main/scala/mesosphere/marathon/core/election/CuratorElectionStream.scala

```diff
@@ -0,0 +1,192 @@
+        val emitterLogic = new CuratorEventEmitter(client, config.zooKeeperLeaderPath, hostPort, sq)
+        emitterLogic.start()
+      } catch {
+        case ex: Throwable =>
+          sq.fail(ex)
+      }
+
+      new CancellableOnce(() => sq.complete())
+    }
+
+    // When the stream closes, we want the last word to be that there is no leader
+    leaderEvents
+      .initialTimeout(config.zooKeeperConnectionTimeout().millis)
+      .concat(Source.single(LeadershipState.Standby(None)))
+  }
+
+  private class CuratorEventEmitter(
+      client: CuratorFramework,
+      zooKeeperLeaderPath: String,
+      hostPort: String, sq: SourceQueueWithComplete[LeadershipState])(implicit actorSystem: ActorSystem) {
+    private lazy val threadExecutor: ExecutorService = Executors.newSingleThreadExecutor()
+    private implicit lazy val singleThreadEC: ExecutionContext = ExecutionContext.fromExecutor(threadExecutor)
+    private lazy val leaderHostPortMetric: Timer = Metrics.timer(ServiceMetric, getClass, "current-leader-host-port")
+    private lazy val latch = new LeaderLatch(client, zooKeeperLeaderPath + "-curator", hostPort)
+
+    @volatile private var _isLeader: Boolean = false
+    @volatile private var _currentLeader: Option[String] = None
+    @volatile private var pollCancellable: Option[Cancellable] = None
+    @volatile private var started = false
+    @volatile private var stopped = false
+
+    private def getCurrentLeader: Option[String] = leaderHostPortMetric.blocking {
+      if (client.getState == CuratorFrameworkState.STOPPED)
+        None
+      else {
+        try {
+          val participant = latch.getLeader
+          if (participant.isLeader) Some(participant.getId) else None
+        } catch {
+          case ex: Throwable =>
+            logger.error("Error while getting current leader", ex)
+            None
+        }
+      }
+    }
+
+    private val listener: LeaderLatchListener = new LeaderLatchListener {
+      override def notLeader(): Unit = {
+        _isLeader = true
+        _currentLeader = getCurrentLeader
+        sq.offer(LeadershipState.Standby(_currentLeader))
+        logger.info(s"Leader defeated. New leader: ${_currentLeader.getOrElse("-")}")
+      }
+
+      override def isLeader(): Unit = {
+        _isLeader = true
+        sq.offer(LeadershipState.ElectedAsLeader)
+      }
+    }
+
+    def start(): Unit = synchronized {
+      assert(!started, "already started")
+      started = true
+      // Register hook such that this logic module will tear itself down if the stream stops for any reason
+      sq.watchCompletion().onComplete { _ => stop() }
+
+      try {
+        latch.addListener(listener, threadExecutor)
+        latch.start()
+
+        // Note: runs in same executor as listener; therefore serial
+        pollCancellable = Some(actorSystem.scheduler.schedule(1.seconds, 1.second){
+          if (!_isLeader) {
+            val nextLeader = getCurrentLeader
```

Comment:
We used to poll in the standby everytime a proxies HTTP endpoint method was hit; perhaps it'd be better to use a watcher, instead? So Zookeeper can push events to us when the leader changes, rather than polls?